lib/hexpm/repository/owners.ex

```diff
@@ -16,36 +16,79 @@ defmodule Hexpm.Repository.Owners do
   def add(package, user, params, audit: audit_data) do
     repository = package.repository
     owners = all(package, user: :emails)
+    organization_owner = Enum.find(owners, & &1.user.organization_id)
+    owner_organization = organization_owner && Repo.get(Organization, organization_owner.user.organization_id)
+
+    repository_access = Organizations.access?(repository.organization, user, "read")
+    organization_access = owner_organization && Organizations.access?(owner_organization, user, "read")
+    new_owner_is_organization? = !!user.organization_id
+
+    cond do
+      !repository.public && !repository_access ->
+        {:error, :not_member}
+
+      # Outside collaborators are not allowed at this time
+      owner_organization && !organization_access ->
+        {:error, :not_member}
+
+      new_owner_is_organization? && Map.get(params, "transfer", false) != true ->
+        {:error, :not_organization_transfer}
+
+      new_owner_is_organization? && Map.get(params, "level", "full") != "full" ->
+        {:error, :organization_level}
+
+      true ->
+        add_owner(package, owners, user, params, audit_data)
+    end
+  end
+
+  defp add_owner(package, owners, user, params, audit_data) do
     owner = Enum.find(owners, &(&1.user_id == user.id))
     owner = owner || %PackageOwner{package_id: package.id, user_id: user.id}
     changeset = PackageOwner.changeset(owner, params)
 
-    if repository.public or Organizations.access?(repository.organization, user, "read") do
-      multi =
-        Multi.new()
-        |> Multi.insert_or_update(:owner, changeset)
-        |> audit(audit_data, "owner.add", fn %{owner: owner} ->
-          {package, owner.level, user}
-        end)
-
-      case Repo.transaction(multi) do
-        {:ok, %{owner: owner}} ->
-          # TODO: Separate email for the affected person
-          owners = Enum.map(owners, & &1.user)
-
-          Emails.owner_added(package, [user | owners], user)
-          |> Mailer.deliver_now_throttled()
-
-          {:ok, %{owner | user: user}}
-
-        {:error, :owner, changeset, _} ->
-          {:error, changeset}
-      end
-    else
-      {:error, :not_member}
+    multi =
+      Multi.new()
+      |> Multi.insert_or_update(:owner, changeset)
+      |> remove_existing_owners(owners, params)
+      |> audit(audit_data, "owner.add", fn %{owner: owner} ->
```

Comment:
What do you think about calling this action `"owner.transfer"`? Otherwise I think we should at least add logs for removal of each owner but I think logging this as `"owner.transfer"` is better :)

Reply:
Very good point.